### PR TITLE
Replace operators with wrapper functions on universal intrinsics backends

### PR DESCRIPTION
**Pull request by WanliZhong**
This PR aims to replace the operators(logic, arithmetic, bit) with wrapper functions(v_add, v_eq, v_and...)

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom
build_image:Custom=simd-emulator
buildworker:Custom=linux-1,linux-4
```

### CODE REVIEW

File: modules/core/include/opencv2/core/hal/intrin.hpp [hunk 1]
```diff
@@ -727,34 +727,11 @@ namespace CV__SIMD_NAMESPACE {
...
-    #define OPENCV_HAL_WRAP_SHIFT_OP(_Tpvec) \
-    inline _Tpvec v_shr(const _Tpvec& a, int n) \
-    { \
-        return a >> n; \
-    } \
-    inline _Tpvec v_shl(const _Tpvec& a, int n) \
-    { \
-        return a << n; \
+        return v_add(v_add(f1, f2), vf...); \
     }
```
Comment (WanliZhong) [inline modules/core/include/opencv2/core/hal/intrin.hpp:733]: This function causes a compilation error, which recursively calls itself. But it should call the `v_add` function from other backends. This function works well but fails on `v_add` function from `intrin_cpp.hpp`.

```
In file included from /home/wanli/workspace/opencv_china/opencv/modules/core/test/test_intrin128.simd.hpp:7,
                 from /home/wanli/workspace/opencv_china/opencv/modules/core/test/test_intrin_emulator.cpp:14:
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin.hpp: In instantiation of ‘cv::hal_EMULATOR_CPP::simd128_cpp::v_uint8 cv::hal_EMULATOR_CPP::simd128_cpp::v_add(const v_uint8&, const v_uint8&, const Args& ...) [with Args = {}; v_uint8 = cv::hal_EMULATOR_CPP::v_reg<unsigned char, 16>]’:
/home/wanli/workspace/opencv_china/opencv/modules/core/test/test_intrin_utils.hpp:491:29:   required from ‘opencv_test::hal::intrin128::opt_EMULATOR_CPP::TheTest<R>& opencv_test::hal::intrin128::opt_EMULATOR_CPP::TheTest<R>::test_addsub() [with R = cv::hal_EMULATOR_CPP::v_reg<unsigned char, 16>]’
/home/wanli/workspace/opencv_china/opencv/modules/core/test/test_intrin_utils.hpp:1923:21:   required from here
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin.hpp:732:21: error: no matching function for call to ‘v_add(cv::hal_EMULATOR_CPP::simd128_cpp::v_uint8)’
  732 |         return v_add(v_add(f1, f2), vf...); \
      |                ~~~~~^~~~~~~~~~~~~~~~~~~~~~
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin.hpp:735:5: note: in expansion of macro ‘OPENCV_HAL_WRAP_BIN_OP_ADDSUB’
  735 |     OPENCV_HAL_WRAP_BIN_OP_ADDSUB(v_uint8)
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin.hpp:731:19: note: candidate: ‘template<class ... Args> cv::hal_EMULATOR_CPP::simd128_cpp::v_uint8 cv::hal_EMULATOR_CPP::simd128_cpp::v_add(const v_uint8&, const v_uint8&, const Args& ...)’
  731 |     inline _Tpvec v_add(const _Tpvec& f1, const _Tpvec& f2, const Args&... vf) { \
      |                   ^~~~~
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin.hpp:735:5: note: in expansion of macro ‘OPENCV_HAL_WRAP_BIN_OP_ADDSUB’
  735 |     OPENCV_HAL_WRAP_BIN_OP_ADDSUB(v_uint8)
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin.hpp:731:19: note:   template argument deduction/substitution failed:
  731 |     inline _Tpvec v_add(const _Tpvec& f1, const _Tpvec& f2, const Args&... vf) { \
      |                   ^~~~~
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin.hpp:735:5: note: in expansion of macro ‘OPENCV_HAL_WRAP_BIN_OP_ADDSUB’
  735 |     OPENCV_HAL_WRAP_BIN_OP_ADDSUB(v_uint8)
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin.hpp:732:21: note:   candidate expects at least 2 arguments, 1 provided
  732 |         return v_add(v_add(f1, f2), vf...); \
      |                ~~~~~^~~~~~~~~~~~~~~~~~~~~~
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin.hpp:735:5: note: in expansion of macro ‘OPENCV_HAL_WRAP_BIN_OP_ADDSUB’
  735 |     OPENCV_HAL_WRAP_BIN_OP_ADDSUB(v_uint8)
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin.hpp:247:
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin_cpp.hpp:576:55: note: candidate: ‘template<class _Tp, int n> cv::hal_EMULATOR_CPP::v_reg<_Tp, n> cv::hal_EMULATOR_CPP::v_add(const v_reg<_Tp, n>&, const v_reg<_Tp, n>&)’
  576 | template<typename _Tp, int n> CV_INLINE v_reg<_Tp, n> v_add(const v_reg<_Tp, n>& a, const v_reg<_Tp, n>& b);
      |                                                       ^~~~~
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin_cpp.hpp:576:55: note:   template argument deduction/substitution failed:
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin.hpp:732:21: note:   candidate expects 2 arguments, 1 provided
  732 |         return v_add(v_add(f1, f2), vf...); \
      |                ~~~~~^~~~~~~~~~~~~~~~~~~~~~
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin.hpp:735:5: note: in expansion of macro ‘OPENCV_HAL_WRAP_BIN_OP_ADDSUB’
  735 |     OPENCV_HAL_WRAP_BIN_OP_ADDSUB(v_uint8)
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core.hpp:52,
                 from /home/wanli/workspace/opencv_china/opencv/modules/ts/include/opencv2/ts.hpp:10,
                 from /home/wanli/workspace/opencv_china/opencv/modules/core/test/test_precomp.hpp:9,
                 from /home/wanli/workspace/opencv_china/opencv/modules/core/test/test_intrin_emulator.cpp:4:
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin_cpp.hpp:647:31: note: candidate: ‘template<int n> cv::hal_EMULATOR_CPP::v_reg<unsigned char, n> cv::hal_EMULATOR_CPP::v_add(const v_reg<unsigned char, n>&, const v_reg<unsigned char, n>&)’
  647 | CV__HAL_INTRIN_IMPL_BIN_OP(+, v_add)
      |                               ^~~~~
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/cvdef.h:79:24: note: in definition of macro ‘__CV_EXPAND’
   79 | #define __CV_EXPAND(x) x
      |                        ^
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin_cpp.hpp:618:13: note: in expansion of macro ‘CV__HAL_INTRIN_IMPL_BIN_OP_’
  618 | __CV_EXPAND(macro_name(uchar, __VA_ARGS__)) \
      |             ^~~~~~~~~~
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin_cpp.hpp:632:1: note: in expansion of macro ‘CV__HAL_INTRIN_EXPAND_WITH_INTEGER_TYPES’
  632 | CV__HAL_INTRIN_EXPAND_WITH_INTEGER_TYPES(macro_name, __VA_ARGS__) \
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin_cpp.hpp:645:50: note: in expansion of macro ‘CV__HAL_INTRIN_EXPAND_WITH_ALL_TYPES’
  645 | #define CV__HAL_INTRIN_IMPL_BIN_OP(bin_op, func) CV__HAL_INTRIN_EXPAND_WITH_ALL_TYPES(CV__HAL_INTRIN_IMPL_BIN_OP_, bin_op, func)
      |                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin_cpp.hpp:647:1: note: in expansion of macro ‘CV__HAL_INTRIN_IMPL_BIN_OP’
  647 | CV__HAL_INTRIN_IMPL_BIN_OP(+, v_add)
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin_cpp.hpp:647:31: note:   template argument deduction/substitution failed:
  647 | CV__HAL_INTRIN_IMPL_BIN_OP(+, v_add)
      |                               ^~~~~
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/cvdef.h:79:24: note: in definition of macro ‘__CV_EXPAND’
   79 | #define __CV_EXPAND(x) x
      |                        ^
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin_cpp.hpp:618:13: note: in expansion of macro ‘CV__HAL_INTRIN_IMPL_BIN_OP_’
  618 | __CV_EXPAND(macro_name(uchar, __VA_ARGS__)) \
      |             ^~~~~~~~~~
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin_cpp.hpp:632:1: note: in expansion of macro ‘CV__HAL_INTRIN_EXPAND_WITH_INTEGER_TYPES’
  632 | CV__HAL_INTRIN_EXPAND_WITH_INTEGER_TYPES(macro_name, __VA_ARGS__) \
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/wanli/workspace/opencv_china/opencv/modules/core/include/opencv2/core/hal/intrin_cpp.hpp:645:50: note: in expansion of macro ‘CV__HAL_INTRIN_EXPAND_WITH_ALL_TYPES’
  645 | #define CV__HAL_INTRIN_IMPL_BIN_OP(bin_op, func) CV__HAL_INTRIN_EXPAND_WITH_ALL_TYPES(CV__HAL_INTRIN_IMPL_BIN_OP_, bin_op, func)
      |                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.................
```

**Comment by vpisarev** [comment]
@WanliZhong, please, check it; it now does not complain on v_add & v_mul, but complains about v_exp etc.

**Comment by asmorkalov** [comment]
@fengyuentau could you check longsoon branch?

**Comment by asmorkalov** [comment]
@mshabunin @opencv-alalek could you take a look?

**Comment by WanliZhong** [comment]
Loonix test results:
**opencv_test_core: **
```
1: [==========] 12030 tests from 281 test cases ran. (231248 ms total)
1: [  PASSED  ] 12023 tests.
1: [  FAILED  ] 7 tests, listed below:
1: [  FAILED  ] Core_InputOutput.filestorage_base64_basic_read_XML
1: [  FAILED  ] Core_InputOutput.filestorage_base64_basic_read_YAML
1: [  FAILED  ] Core_InputOutput.filestorage_base64_basic_read_JSON
1: [  FAILED  ] Core_InputOutput.filestorage_base64_basic_rw_XML
1: [  FAILED  ] Core_InputOutput.filestorage_base64_basic_rw_YAML
1: [  FAILED  ] Core_InputOutput.filestorage_base64_basic_rw_JSON
1: [  FAILED  ] Core_globbing.accuracy
```

**Comment by WanliZhong** [comment]
@asmorkalov Longsoon tests are all passed. (GUI test failed is not related to this PR, because it's server)
Screenshot 2024-09-12 at 12 36 35

**Comment by asmorkalov** [comment]
cc @hanliutong

File: modules/core/include/opencv2/core/hal/intrin_wasm.hpp [hunk 2]
```diff
@@ -1694,7 +1687,7 @@ inline unsigned v_reduce_sad(const v_uint8x16& a, const v_uint8x16& b)
     v_expand(v_absdiff(a, b), l16, h16);
     v_expand(l16, l16_l32, l16_h32);
     v_expand(h16, h16_l32, h16_h32);
-    return v_reduce_sum(l16_l32+l16_h32+h16_l32+h16_h32);
```
Comment (asmorkalov) [inline modules/core/include/opencv2/core/hal/intrin_wasm.hpp:1697]: you can reduce amount of operations here like (a+b) + (c+d).

File: modules/core/include/opencv2/core/hal/intrin_wasm.hpp [hunk 3]
```diff
@@ -1703,19 +1696,19 @@ inline unsigned v_reduce_sad(const v_int8x16& a, const v_int8x16& b)
...
```
Comment (asmorkalov) [inline modules/core/include/opencv2/core/hal/intrin_wasm.hpp:1706]: and here.

**Comment by WanliZhong** [comment]
The default test error about `v_exp` should be fixed in https://github.com/opencv/opencv/pull/25891, I have included this PR and it passed the test but not in this PR.

```
In file included from /build/precommit_custom_linux/4.x/opencv/modules/core/include/opencv2/core/hal/intrin.hpp:961:0,
                 from /build/precommit_custom_linux/4.x/opencv/modules/core/src/precomp.hpp:88,
                 from /build/precommit_custom_linux/build/modules/core/mathfuncs_core.avx2.cpp:2:
/build/precommit_custom_linux/4.x/opencv/modules/core/include/opencv2/core/hal/intrin_math.hpp: In function 'cv::hal_EMULATOR_CPP::simd256::v_float32 cv::hal_EMULATOR_CPP::simd256::v_erf(const v_float32&)':
/build/precommit_custom_linux/4.x/opencv/modules/core/include/opencv2/core/hal/intrin_math.hpp:453:40: error: no matching function for call to 'v_exp(cv::hal_EMULATOR_CPP::simd256::v_float32&)'
         v_float32 exp = v_exp(neg_pow_2);
                                        ^
 ```

**Pull request merged**